src/tests/GfxBatchHbaoTest.cpp

```diff
@@ -0,0 +1,352 @@
+     esp::gfx_batch::HbaoConfiguration{}.flags(),
+     esp::gfx_batch::HbaoConfiguration{}.intensity(),
+     esp::gfx_batch::HbaoConfiguration{}.radius(),
+     esp::gfx_batch::HbaoConfiguration{}.blurSharpness(), 0.0f, 0.0f},
+    {"classic, orthographic, AO special blur",
+     "hbao_tests/van-gogh-room.hbao-classic-sblur-ortho.png", true, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags() |
+         esp::gfx_batch::HbaoFlag::UseAoSpecialBlur,
+     esp::gfx_batch::HbaoConfiguration{}.intensity(),
+     esp::gfx_batch::HbaoConfiguration{}.radius(),
+     esp::gfx_batch::HbaoConfiguration{}.blurSharpness(), 2.0f, 2.0f},
+    {"classic, orthographic, strong effect",
+     "hbao_tests/van-gogh-room.hbao-classic-strong-ortho.png", true, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags(), 2.0f, 1.5f, 10.0f, 0.0f,
+     0.0f},
+
+    {"cache-aware, orthographic, defaults",
+     "hbao_tests/van-gogh-room.hbao-cache-ortho.png", false, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags(),
+     esp::gfx_batch::HbaoConfiguration{}.intensity(),
+     esp::gfx_batch::HbaoConfiguration{}.radius(),
+     esp::gfx_batch::HbaoConfiguration{}.blurSharpness(), 0.0f, 0.0f},
+    {"cache-aware, orthographic, AO special blur",
+     "hbao_tests/van-gogh-room.hbao-cache-sblur-ortho.png", false, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags() |
+         esp::gfx_batch::HbaoFlag::UseAoSpecialBlur,
+     esp::gfx_batch::HbaoConfiguration{}.intensity(),
+     esp::gfx_batch::HbaoConfiguration{}.radius(),
+     esp::gfx_batch::HbaoConfiguration{}.blurSharpness(), 2.0f, 2.0f},
+    {"cache-aware, orthographic, strong effect",
+     "hbao_tests/van-gogh-room.hbao-cache-strong-ortho.png", false, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags(), 2.0f, 1.5f, 10.0f, 0.0f,
+     0.0f},
+    {"cache-aware, orthographic, layered with image load store",
+     "hbao_tests/van-gogh-room.hbao-cache-layered-ortho.png", false, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags() |
+         esp::gfx_batch::HbaoFlag::LayeredImageLoadStore,
+     esp::gfx_batch::HbaoConfiguration{}.intensity(),
+     esp::gfx_batch::HbaoConfiguration{}.radius(),
+     esp::gfx_batch::HbaoConfiguration{}.blurSharpness(), 4.5f, 0.11f},
+    {"cache-aware, orthographic, layered with geometry shader",
+     "hbao_tests/van-gogh-room.hbao-cache-geom-ortho.png", false, false,
+     esp::gfx_batch::HbaoConfiguration{}.flags() |
+         esp::gfx_batch::HbaoFlag::LayeredGeometryShader,
+     esp::gfx_batch::HbaoConfiguration{}.intensity(),
+     esp::gfx_batch::HbaoConfiguration{}.radius(),
+     esp::gfx_batch::HbaoConfiguration{}.blurSharpness(), 4.5f, 0.11f},
+};
+
+GfxBatchHbaoTest::GfxBatchHbaoTest() {
+  addTests({&GfxBatchHbaoTest::generateTestData});
+
+  addInstancedTests({&GfxBatchHbaoTest::test},
+                    Cr::Containers::arraySize(TestData));
+}
+
+constexpr Mn::Vector2i Size{256, 256};
```

Comment:
Got an idea for easy verification that the aspect ratio handling is implemented correctly: change this to something non-square (that still reasonably covers the scene, of course) and add a second test variant (`testRotated()`?) that uses `Size.flipped()` instead of `Size` and loads the images rotated by 90°:

```cpp
/* This rotates the depth image by 90° */
Cr::Containers::Array<char> rotatedDepth{Cr::NoInit, depth->data.size()};
Cr::Utility::copy(depth->pixels().flipped<1>().transposed<0, 1>(), 
    Cr::Containers::StridedArrayView2D<char>{rotatedDepth, {std::size_t(Size.y()), std::size_t(Size.x())}});

Mn::GL::Texture2D inputDepthTexture;
inputDepthTexture
    .setStorage(1, Mn::GL::TextureFormat::DepthComponent32F, Size.flipped())
    .setSubImage(0, {}, Mn::ImageView2D{depth->format(), Size.flipped(), rotatedDepth});
```

Similarly for the color image. The final comparison would be to the original image, but with the pixels rotated back:

```cpp
    CORRADE_COMPARE_WITH(
            output.read({{}, Size}, {Mn::PixelFormat::RGBA8Unorm})
                .pixels<Color4ub>().transposed<1, 0>().flipped<1>(),
            Cr::Utility::Path::join(TEST_ASSETS, data.filename),
            (Mn::DebugTools::CompareImageToFile{data.maxThreshold, data.meanThreshold}));
```

If everything is correct in the implementation, the test should pass without any difference compared to the non-rotated case.

(Details about strided array views and their dimension flipping and transposition are in [this blog post](https://blog.magnum.graphics/backstage/multidimensional-strided-array-views/).)

Reply:
Put these tests in and they are exposing a bias in the algorithm that seems like it exists in the shader, so I have marked them Expect to Fail and will investigate this further. The effect is present and visually pleasing, however, so I don't think this should hold up the PR.